miner/Miner.hs

```diff
@@ -0,0 +1,285 @@
+import Crypto.Hash.Algorithms (SHA512t_256)
+import Crypto.Hash.IO
+import qualified Data.ByteArray as BA
+import Data.Bytes.Put
+import qualified Data.ByteString as B
+import Data.HashMap.Strict (HashMap)
+import qualified Data.HashMap.Strict as HM
+import qualified Data.Text as T
+import Data.Tuple.Strict (T2(..))
+import Foreign.Marshal.Alloc (allocaBytes)
+import qualified Network.Wai.Handler.Warp as W
+import Options.Applicative
+import Servant.API
+import Servant.Server (Application, Server, serve)
+import Servant.Swagger (toSwagger)
+import Servant.Swagger.UI (SwaggerSchemaUI, swaggerSchemaUIServer)
+
+-- internal modules
+import Chainweb.BlockHeader
+import Chainweb.Difficulty (HashTarget, encodeHashTarget)
+import Chainweb.RestAPI.Orphans ()
+import Chainweb.Time (Micros, Time, encodeTimeToWord64, getCurrentTimeIntegral)
+import Chainweb.Utils (int, runGet)
+import Chainweb.Version (ChainId, ChainwebVersion(..), chainwebVersionFromText)
+import Miner.Types ()
+
+---
+
+--------------------------------------------------------------------------------
+-- Servant
+
+type MinerAPI = "submit" :> ReqBody '[JSON] BlockHeader :> Post '[JSON] ()
+    :<|> "poll" :> Capture "chainid" ChainId
+                :> Capture "blockheight" BlockHeight
+                :> Get '[JSON] (Maybe BlockHeader)
+
+type API = MinerAPI :<|> SwaggerSchemaUI "help" "swagger.json"
+
+server :: Env -> Server API
+server e = (liftIO . submit e :<|> (\cid h -> liftIO $ poll e cid h))
+    :<|> swaggerSchemaUIServer (toSwagger (Proxy :: Proxy MinerAPI))
+
+app :: Env -> Application
+app = serve (Proxy :: Proxy API) . server
+
+--------------------------------------------------------------------------------
+-- CLI
+
+type ResultMap = HashMap (T2 ChainId BlockHeight) BlockHeader
+
+data Env = Env
+    { work :: TMVar BlockHeader
+    , results :: TVar ResultMap
+    , cores :: Word16
+    , version :: ChainwebVersion
+    , port :: Int
+    }
+
+pEnv :: TMVar BlockHeader -> TVar ResultMap -> Parser Env
+pEnv tbh thm = Env tbh thm <$> pCores <*> pVersion <*> pPort
+
+pCores :: Parser Word16
+pCores = option auto
+    (long "cores" <> metavar "COUNT" <> value 1
+     <> help "Number of CPU cores to use (default: 1)")
+
+pVersion :: Parser ChainwebVersion
+pVersion = option cver
+    (long "version" <> metavar "VERSION"
+     <> value Testnet01
+     <> help "Chainweb Network Version (default: testnet01)")
+  where
+    cver :: ReadM ChainwebVersion
+    cver = eitherReader $ \s ->
+        note "Illegal ChainwebVersion" . chainwebVersionFromText $ T.pack s
+
+pPort :: Parser Int
+pPort = option auto
+    (long "port" <> metavar "PORT" <> value 8081
+     <> help "Port on which to run the miner (default: 8081)")
+
+--------------------------------------------------------------------------------
+-- Work
+
+main :: IO ()
+main = do
+    env <- (opts <$> newEmptyTMVarIO <*> newTVarIO mempty) >>= execParser
```

Comment:
Let's use the same configuration system as for Chainweb-node. One motivation behind `configuration-tools` was that it makes it easy to use the same consistent configuration across different components of micro services. If we don't like it, we can use something else, but it should be consistent for all chainweb components. 

In particular when we compile the components into a single standalone node (for desktop users) we want the configuration file format and command line be consistent with the stand-alone components.

Reply:
My thought here was that `configuration-tools` was overkill: we only have three possible flags. Using raw `optparse-applicative` kept things simple.